src/Interpreters/Aggregator.cpp

```diff
@@ -1006,7 +1053,21 @@ void NO_INLINE Aggregator::executeImplBatch(
         /// For all rows.
         AggregateDataPtr place = aggregates_pool->alloc(0);
         for (size_t i = row_begin; i < row_end; ++i)
+        {
+            if constexpr (prefetch && HasPrefetchMemberFunc<decltype(method.data), KeyHolder>)
+            {
+                if (i == row_begin + prefetching.iterationsToMeasure())
+                    prefetch_look_ahead = prefetching.calcPrefetchLookAhead();
+
+                if (i + prefetch_look_ahead < row_end)
+                {
+                    auto && key_holder = state.getKeyHolder(i + prefetch_look_ahead, *aggregates_pool);
+                    method.data.prefetch(std::move(key_holder));
```

Comment:
Hi @nickitat , it seems that the prefetching operations add some overhead in double serialization and hashing.
Here, getKeyHolder would serialize the columns into a key; method.data.prefetch() would compute the hash value of the key. In the next line, state.emplaceKey duplicates these two steps. 
Would it be better to cache the serialized key and hash value? Thanks!

Reply:
for this reason we have special trait in HashMethod-s that tells, if these operations are cheap:

``` cpp
const bool prefetch = Method::State::has_cheap_key_calculation && params.enable_prefetch
             && (method.data.getBufferSizeInBytes() > min_bytes_for_prefetch);
```

another consideration is that prefetching gives equal speedup regardless of the key complexity and the more costly to compute keys the less noticeable will be perf gain from prefetching, so we not really loose much if we just disable the optimisation for group by with complex keys.

that being said cache keys makes perfect sense. it wasn't implemented mostly because I didn't want to introduce some amount of fairly ugly code. it would be required because for example some key holders are not copiable (e.g. `ArenaKeyHolder`)